### PR TITLE
refactor!: remove algo in orchestrator task

### PR DESCRIPTION
**Pull request by guilhem-barthes**
## Description

Remove the embedded algo in the orchestrator.

This will align the compute task gRPC resource to the global philosophy of depdendent resources, where we only pass references to the resources instead of embedding them.

## Companion PR

- https://github.com/Substra/substra-backend/pull/474

## How has this been tested?

Local substra test has been run locally and full e2e test suite has been run in CI.

## Checklist

- [x] [changelog](../CHANGELOG.md) was updated with notable changes

### CODE REVIEW

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feature/remove-algo-in-orchestrator-task

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by Owlfred** [comment]
```
Usage: /e2e [options] [help]

/e2e may appear anywhere as long as it is on its own line

Options:
  --refs <value>                                         Extra refs (branch or tag) with format REPO=GIT_REF,REPO=GIT_REF.
  Supported repositories: substra-hlf-k8s, orchestrator, substra-backend, substra-frontend, substra-tools, substrafl, substra, substra-tests, connect-ci.
  Example: /e2e --refs substra-backend=some_branch,orchestrator=some_tag (default: {})
  --tests-to-run, --tests <tests-to-run>                 Comma-separated list of tests to run. Valid options: sdk,substrafl,frontend or NONE. (default: "sdk")
  --benchmarks-to-run, --benchmarks <benchmarks-to-run>  Comma-separated list of workflows tests to run. Valid options: mnist, camelyon or NONE. (default: "NONE")
  --orchestrator-mode, --mode <orchestrator-mode>        Comma-separated list of orchestrator modes to run tests for. Valid options: standalone,distributed (default: "standalone,distributed")
  -h, --help                                             display help for command
```

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feature/remove-algo-in-orchestrator-task --mode standalone

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: server/standalone/migration/000045_remove_algo_from_compute_task.up.sql [hunk 1]
```diff
@@ -20,3 +20,9 @@ FROM compute_tasks t
     FROM compute_task_parents
     GROUP BY child_task_key
 ) p ON p.child_task_key = t.key;
+
+
+UPDATE events
+/* Only update algo.key. Don't update the other algo.* fields (too complex) */
```
Comment (mblottiere) [inline server/standalone/migration/000045_remove_algo_from_compute_task.up.sql:26]: We only need `algo_key` so this is enough 🤔

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feature/remove-algo-in-orchestrator-task --mode standalone

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feature/remove-algo-in-orchestrator-task --mode standalone

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feature/remove-algo-in-orchestrator-task

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feature/remove-algo-in-orchestrator-task

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: server/standalone/migration/000045_remove_algo_from_compute_task.down.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+/* intentionnaly empty */
```
Comment (thibowk) [inline server/standalone/migration/000045_remove_algo_from_compute_task.down.sql:1]:
```suggestion
/* intentionally empty */
```

**Review by thibowk — COMMENTED** [review]
Thanks for the updates!

Overall, it looks good to me. Nice job!

One tiny question: is there anything preventing us from providing a down migration? If no, I guess we should do so.

File: CHANGELOG.md [hunk 3]
```diff
@@ -20,6 +21,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Worker field on NewAggregateTrainTaskData, use NewComputeTask.Worker field instead
 - DisableModel rpc in Model service, use DisableOutput in ComputeTask service instead
 - CanDisableModel rpc in Model service.
+- (BREAKING): Removed fields related to algo (except `algo_key`) in view `expanded_compute_tasks`
```
[the remark this answers is not in the document]
Comment (guilhem-barthes) [reply CHANGELOG.md]: Addressed in 7f15df585446a0937b2dfcf87bc4a7b88a699b60

File: server/standalone/migration/000045_remove_algo_from_compute_task.up.sql (hunk 1 above)
Comment (guilhem-barthes) [reply server/standalone/migration/000045_remove_algo_from_compute_task.up.sql]: Addressed in 7f15df585446a0937b2dfcf87bc4a7b88a699b60

**Comment by guilhem-barthes** [comment]
To create the down migration, it would require to join on the algo table and add algo data to the compute_task & the serialized algo in event, which is doable.

**Comment by guilhem-barthes** [comment]
> Thanks for the updates!
> 
> Overall, it looks good to me. Nice job!
> 
> One tiny question: is there anything preventing us from providing a down migration? If no, I guess we should do so.

This has been addressed in [commit 4bb3ed1004ef8345683722f713c1c1748776fb32](https://github.com/Substra/orchestrator/commit/4bb3ed1004ef8345683722f713c1c1748776fb32)

File: server/standalone/migration/000046_remove_algo_from_compute_task.down.sql [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+         LEFT JOIN expanded_algos a ON a.key = t.algo_key
+         LEFT JOIN (
+    SELECT child_task_key, JSONB_AGG(parent_task_key ORDER BY position) AS parent_task_keys
+    FROM compute_task_parents
+    GROUP BY child_task_key
+) p ON p.child_task_key = t.key;
+
+UPDATE events AS e
+SET asset = jsonb_set(asset, '{algo}', jsonb_build_object('key', asset->>'algoKey'
+                                        , 'algo_name', a.name
```
Comment (thibowk) [inline server/standalone/migration/000046_remove_algo_from_compute_task.down.sql:36]: In the current version, an example of a compute task representation stored in an event is :

```json
{
   "key":"00798f1a-0098-4430-beae-65dd8c7c00c5",
   "algo":{
      "key":"50f2fae2-e8b9-456a-b218-a41c72dbef88",
      "name":"Algo test",
      "owner":"MyOrg1MSP",
      "inputs":{
         "model":{
            "kind":"ASSET_MODEL",
            "multiple":true,
            "optional":true
         },
         "opener":{
            "kind":"ASSET_DATA_MANAGER"
         },
         "datasamples":{
            "kind":"ASSET_DATA_SAMPLE",
            "multiple":true
         }
      },
      "outputs":{
         "model":{
            "kind":"ASSET_MODEL"
         }
      },
      "algorithm":{
         "checksum":"1d55e9c55fa7ad6b6a49ad79da897d58be7ce8b76f92ced4c20f361ba3a0af6e",
         "storageAddress":"http://somewhere.local/algo/72d8aa85-0901-444b-ada4-e4fafab6ac04"
      },
      "description":{
         "checksum":"1d55e9c55fa7ad6b6a49ad79da897d58be7ce8b76f92ced4c20f361ba3a0af6e",
         "storageAddress":"http://somewhere.local/desc/3cf45ebc-fd51-4744-8504-36e9267ef06d"
      },
      "permissions":{
         "process":{
            "public":true,
            "authorizedIds":[
               "MyOrg1MSP"
            ]
         },
         "download":{
            "public":true,
            "authorizedIds":[
               "MyOrg1MSP"
            ]
         }
      },
      "creationDate":"2022-09-26T13:46:54.337609Z"
   },
   "rank":1,
   "owner":"MyOrg1MSP",
   "train":{
      "dataManagerKey":"132bd2b5-5844-43df-8a51-21b40550a784",
      "dataSampleKeys":[
         "a0d553bd-583a-4cb3-83b7-1f75ee92f3dc"
      ]
   },
   "inputs":[
      {
         "assetKey":"132bd2b5-5844-43df-8a51-21b40550a784",
         "identifier":"opener"
      },
      {
         "assetKey":"a0d553bd-583a-4cb3-83b7-1f75ee92f3dc",
         "identifier":"datasamples"
      },
      {
         "identifier":"model",
         "parentTaskOutput":{
            "parentTaskKey":"0d9c0c67-c284-48b0-8c6f-d1ece817b134",
            "outputIdentifier":"model"
         }
      }
   ],
   "status":"STATUS_WAITING",
   "worker":"MyOrg1MSP",
   "outputs":{
      "model":{
         "permissions":{
            "process":{
               "public":true,
               "authorizedIds":[
                  "MyOrg1MSP"
               ]
            },
            "download":{
               "public":true,
               "authorizedIds":[
                  "MyOrg1MSP"
               ]
            }
         }
      }
   },
   "category":"TASK_TRAIN",
   "creationDate":"2022-09-26T13:46:54.396473Z",
   "computePlanKey":"253b2da2-0e58-4d6b-89e8-30e22d3142d6",
   "logsPermission":{
      "public":true,
      "authorizedIds":[
         "MyOrg1MSP"
      ]
   },
   "parentTaskKeys":[
      "0d9c0c67-c284-48b0-8c6f-d1ece817b134"
   ]
}
```

So it seems that the `algo_` prefix should not be used in the field names. Also, description and algorithm fields are not properly structured. Inputs and outputs seem to be missing as well.

Note: the following helpers functions could be useful https://github.com/Substra/orchestrator/blob/main/server/standalone/migration/000012_add_algo_helpers.up.sql. `build_algo_jsonb` could be updated to be used.

**Review by mblottiere — APPROVED** [review]
LGTM, nice work 👍

File: server/standalone/migration/000046_remove_algo_from_compute_task.down.sql (hunk 4 above)
Comment (guilhem-barthes) [reply server/standalone/migration/000046_remove_algo_from_compute_task.down.sql]: Addressed on https://github.com/Substra/orchestrator/commit/de0dc437bbb25b3dffa327957ec20d5d83c3a4f9

**Review by AurelienGasser — APPROVED** [review]
Nice work :rocket:

File: lib/service/computetask_test.go [hunk 5]
```diff
@@ -450,13 +456,15 @@ func TestRegisterCompositeTaskWithCompositeParents(t *testing.T) {
 	ps.On("CanProcess", algo.Permissions, "testOwner").Once().Return(true)
 
 	// Parent processing check -> requester is the task worker, so the datamanager owner here
+	as.On("GetAlgo", parent1.AlgoKey).Once().Return(algoParent1, nil)
 	ps.On("CanProcess", parent1.Outputs["local"].Permissions, dataManager.Owner).Once().Return(true)
+	as.On("GetAlgo", parent2.AlgoKey).Once().Return(algoParent2, nil)
```
Comment (AlexandrePicosson) [inline lib/service/computetask_test.go:461]: This check is missing at the end of the test function:
```
as.AssertExpectations(t)
```

File: lib/service/model_test.go [hunk 6]
```diff
@@ -157,11 +157,13 @@ func TestRegisterModelWrongPermissions(t *testing.T) {
 }
 
 func TestRegisterTrainModel(t *testing.T) {
+	as := new(MockAlgoAPI)
```
Comment (AlexandrePicosson) [inline lib/service/model_test.go:160]: There is a missing `as.AssertExpectations(t)` at the end of the test function, and all other functions where this service was added.

File: lib/service/performance.go [hunk 7]
```diff
@@ -74,7 +74,12 @@ func (s *PerformanceService) RegisterPerformance(newPerf *asset.NewPerformance,
 	if _, ok := task.Outputs[newPerf.ComputeTaskOutputIdentifier]; !ok {
 		return nil, errors.NewMissingTaskOutput(task.Key, newPerf.ComputeTaskOutputIdentifier)
 	}
-	algoOutput, ok := task.Algo.Outputs[newPerf.ComputeTaskOutputIdentifier]
+	algoTask, err := s.GetAlgoService().GetAlgo(task.AlgoKey)
```
Comment (AlexandrePicosson) [inline lib/service/performance.go:77]: Here you retrieve the same Algo twice, L69 and L77, you could retrieve it once L69, validate that the Algo from L69 have the same key as `task.AlgoKey` then, reuse the Algo from L69 for all the other checks.

Comment (guilhem-barthes) [reply lib/service/performance.go:77]: So we can be sure this is the same algo that is represented by task.AlgoKey and newPerf.MetricKey? If so I will also adapt the tests.

Comment (AlexandrePicosson) [reply lib/service/performance.go:77]: It should be the same, yes if that's not the case you should return an error.

File: lib/service/computetask_test.go (hunk 5 above)
Comment (guilhem-barthes) [reply lib/service/computetask_test.go:461]: So the previous `as.On("GetAlgo", newTask.AlgoKey).Once().Return(algo, nil)` L455 was not doing anything, right?

Comment (AlexandrePicosson) [reply lib/service/computetask_test.go:461]: Not sure what you mean by that but if no one called `GetAlgo` anymore we would not notice it as the mock service does not validate whether it was called or not.

File: lib/asset/computetask.proto [hunk 8]
```diff
@@ -57,9 +56,12 @@ message NewComputeTaskOutput {
 // ComputeTask is a computation step in a ComputePlan.
 // It was previously called XXXtuple: Traintuple, CompositeTraintuple, etc
 message ComputeTask {
+  reserved 3;
+  reserved "algo";
+
   string key = 1;
   ComputeTaskCategory category = 2;
-  Algo algo = 3;
+  string algo_key = 20;
```
[the remark this answers is not in the document]
Comment (guilhem-barthes) [reply lib/asset/computetask.proto]: Addressed in https://github.com/Substra/orchestrator/commit/e92a5c1ca0cff977f9f245483fb73b1f4bfd4b43

File: lib/service/model_test.go (hunk 6 above)
Comment (guilhem-barthes) [reply lib/service/model_test.go:160]: Addressed in b43db3e9ea66bcd5ffc286a907f82a7d210d712f

File: lib/service/computetask_test.go (hunk 5 above)
Comment (guilhem-barthes) [reply lib/service/computetask_test.go:461]: Addressed in b43db3e9ea66bcd5ffc286a907f82a7d210d712f

File: CHANGELOG.md [hunk 9]
```diff
@@ -5,6 +5,16 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
...
+**WARNING**: Some migrations in this version are destructive once applied you will not be able to restore algo categories or algo in serialized events.
+
+### Changed
+- (BREAKING): Replaced `algo` by `algo_key` in ComputeTask
+
+### Fixed
+
+### Added
+
+### Removed
```
[the remark this answers is not in the document]
Comment (guilhem-barthes) [reply CHANGELOG.md]: Addressed in 090226080863db563e7f2e7c9b9812cc342a3e3e

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feature/remove-algo-in-orchestrator-task

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

File: server/standalone/migration/000047_remove_algo_from_compute_task.up.sql [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+         LEFT JOIN (
+    SELECT child_task_key, JSONB_AGG(parent_task_key) AS parent_task_keys
+    FROM compute_task_parents
+    GROUP BY child_task_key
+) p ON p.child_task_key = t.key;
+
+
+UPDATE events
+SET asset = jsonb_set(asset, '{algoKey}', asset->'algo'->'key') #- '{algo}'
+WHERE asset_kind = 'ASSET_COMPUTE_TASK';
```
Comment (AurelienGasser) [inline server/standalone/migration/000047_remove_algo_from_compute_task.up.sql:27]: you should probably add another condition here to only update records which don't already have an algoKey field, to make the migration idempotent.

**Comment by guilhem-barthes** [comment]
/e2e --refs substra-backend=feature/remove-algo-in-orchestrator-task

**Comment by Owlfred** [comment]
End to end tests: :hourglass_flowing_sand: **STARTING**

**Pull request merged**